src/arch/host/include/arch/atomic.h

```diff
@@ -32,6 +32,27 @@ static inline void arch_atomic_init(atomic_t *a, int32_t value)
 	arch_atomic_set(a, value);
 }
 
+#if defined __XCC__
+static inline int32_t arch_atomic_add(atomic_t *a, int32_t value)
+{
+	int32_t result, current;
+
+	current = arch_atomic_read(a);
+	result = current + value;
```

Comment:
Signed overflow is undefined behavior
https://wiki.sei.cmu.edu/confluence/display/c/INT08-C.+Verify+that+all+integer+values+are+in+range
Internal Intel rule C.IA.1
http://blog.llvm.org/2011/05/what-every-c-programmer-should-know_14.html

Reply:
I didn't re-invent the atomic operations, these are copy from arch/xtensa/include/arch/atomic.h. There's also some inline assembler version but they don't look more atomic than this and also use the signed add.
 https://github.com/thesofproject/sof/blob/d7da25fc2537627e16082c0e72e9a9246c92e682/src/arch/xtensa/include/arch/atomic.h#L122